### PR TITLE
Mostly working decoder and moonlight

### PR DESCRIPTION
**Pull request by GavinDarkglider**
There are some issues with the decoder crashing retroarch on video stream close. This generally isnt an issue in lakka because retroarch restarts on close anyway. Will look into a proper fix for that later. For now, it works.

### CODE REVIEW

**Comment by GavinDarkglider** [comment]
Dont merge this yet.

File: projects/L4T/packages/jetson-ffmpeg/patches/CMakeList.patch [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+ target_link_libraries(nvmpi PRIVATE ${CMAKE_THREAD_LIBS_INIT} ${LIB_NVBUF}  ${LIB_V4L2})
++if (LAKKA_API_PATH)
++  target_include_directories(nvmpi PRIVATE ${LAKKA_API_PATH}/usr/include)
++endif ()
++
+ target_include_directories(nvmpi PRIVATE /usr/src/jetson_multimedia_api/include)
+ target_include_directories(nvmpi PRIVATE /usr/local/cuda/include)
+ target_include_directories(nvmpi_static PRIVATE /usr/src/jetson_multimedia_api/include)
+diff -Naur jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef/CMakeLists.txt.orig jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef-2/CMakeLists.txt.orig
+--- jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef/CMakeLists.txt.orig	1970-01-01 00:00:00.000000000 +0000
```
Comment (ToKe79) [inline projects/L4T/packages/jetson-ffmpeg/patches/CMakeList.patch:87]: is this hunk needed?

File: projects/L4T/packages/jetson-ffmpeg/patches/CMakeList.patch [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
++	    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
++	    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
++	        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
++install(FILES ${CMAKE_BINARY_DIR}/nvmpi.pc
++	    DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig)
++install(FILES ${CMAKE_BINARY_DIR}/nvmpi.pc
++	    DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
+diff -Naur jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef/CMakeLists.txt.rej jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef-2/CMakeLists.txt.rej
+--- jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef/CMakeLists.txt.rej	1970-01-01 00:00:00.000000000 +0000
++++ jetson-ffmpeg-20067187641389ba309bd3ca51933718b6b475ef-2/CMakeLists.txt.rej	2021-05-30 17:50:03.585841415 +0000
```
Comment (ToKe79) [inline projects/L4T/packages/jetson-ffmpeg/patches/CMakeList.patch:161]: is this required?

File: projects/L4T/devices/Switch/packages/moonlight/package.mk [hunk 3]
```diff
@@ -0,0 +1,43 @@
+################################################################################
```
Comment (ToKe79) [inline projects/L4T/devices/Switch/packages/moonlight/package.mk:1]: keep original moonlight package, rename this package to e.g. `moonlight-libretro` or `lr-moonlight` or similar (to avoid duplicate package name) and adapt depends / patch names as needed

Comment (ToKe79) [reply projects/L4T/devices/Switch/packages/moonlight/package.mk:1]: also this package can be moved to the proper location for libretro-cores - `packages/libretro`

File: projects/L4T/packages/tegra-bsp/package.mk [hunk 4]
```diff
@@ -19,14 +19,14 @@
 ################################################################################
 
 PKG_NAME="tegra-bsp"
-if [ ! "$DEVICE" = "Switch" ]; then
+if [ ! $DEVICE == "Switch" ]; then
```
Comment (ToKe79) [inline projects/L4T/packages/tegra-bsp/package.mk:22]: double quotes and single equal were OK

File: projects/L4T/packages/tegra-bsp/package.mk [hunk 5]
```diff
@@ -123,21 +154,32 @@ fi
...
+  rm libgbm.so.1
+  cp libnvgbm.so libgbm.so.1
+  #cp libnvgbm.so libgbm.so
+  cp libnvidia-egl-wayland.so libvnidia-egl-wayland.so.1
+  rm libdrm.so.2
+  cp libdrm_nvdc.so libdrm.so.2
   cd $PWD
-  if [ "$DEVICE" = "Switch" ]; then
+
+  if [ "$DEVICE"=="Switch" ]; then
```
Comment (ToKe79) [inline projects/L4T/packages/tegra-bsp/package.mk:177]: single equal + spaces around equal

File: projects/L4T/packages/tegra-bsp/package.mk [hunk 6]
```diff
@@ -123,21 +154,32 @@ fi
...
     #Audio Fix Service
     cp -Pv $PKG_DIR/assets/alsa-fix.service $INSTALL/usr/lib/systemd/system/
     mkdir -p $INSTALL/usr/lib/systemd/system/multi-user.target.wants
     ln -s $INSTALL/usr/lib/systemd/system/alsa-fix.service $INSTALL/usr/lib/systemd/system/multi-user.target.wants/alsa-fix.service
-    if [ "$DISPLAYSERVER" = "x11" ]; then
+    if [ $DISPLAYSERVER="x11" ]; then
```
Comment (ToKe79) [inline projects/L4T/packages/tegra-bsp/package.mk:182]: revert this change

File: projects/L4T/packages/xorg-server/package.mk [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  XORG_COMPOSITE="--disable-composite"
+fi
+
+if [ ! "$OPENGL" = "no" ]; then
+  PKG_DEPENDS_TARGET="$PKG_DEPENDS_TARGET $OPENGL libepoxy"
+  if [ ! "$PROJECT" = "L4T" ]; then 
+    XORG_MESA="--enable-glx --enable-dri --enable-glamor"
+  else  
+    XORG_MESA="--enable-glx --enable-dri --disable-glamor"
+  fi
```
Comment (ToKe79) [inline projects/L4T/packages/xorg-server/package.mk:31]: since the package in under the project L4T, are these tests necessary?
adding the desired options directly to the configure options should be OK.

File: projects/L4T/devices/Switch/packages/moonlight/package.mk [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+#
+#  You should have received a copy of the GNU General Public License
+#  along with OpenELEC.tv; see the file COPYING.  If not, write to
+#  the Free Software Foundation, 51 Franklin Street, Suite 500, Boston, MA 02110, USA.
+#  http://www.gnu.org/copyleft/gpl.html
+################################################################################
+
+PKG_NAME="moonlight"
+PKG_VERSION="20d2d77"
+PKG_ARCH="any"
```
Comment (plaidman) [inline projects/L4T/devices/Switch/packages/moonlight/package.mk:23]: if we're planning on moving this to the main packages/libretro/ folder, this ARCH should be set to only 'aarch64'

File: projects/L4T/options [hunk 9]
```diff
@@ -99,6 +99,7 @@
 
   # override some playlist core assignation
     RA_PLAYLIST_CORES="${RA_PLAYLIST_CORES//parallel_n64/mupen64plus}"
+    #LIBRETRO_CORES="${LIBRETRO_CORES// desmume /}"
```
Comment (plaidman) [inline projects/L4T/options:102]: we can probably kill this line

File: projects/L4T/packages/libglvnd/package.mk [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+#    # Remove old symlinks to GLVND libGL.so.1.7.0
+#    safe_remove              ${INSTALL}/usr/lib/libGL.so
+#    safe_remove              ${INSTALL}/usr/lib/libGL.so.1
+#    # Create new symlinks to /var/lib/libGL.so
+#    ln -sf libGL.so.1        ${INSTALL}/usr/lib/libGL.so
+#    #ln -sf /var/lib/libGL.so ${INSTALL}/usr/lib/libGL.so.1
+#    # Create new symlink to GLVND libGL.so.1.7.0
+#    ln -sf libGL.so.1.7.0    ${INSTALL}/usr/lib/libGL_glvnd.so.1
+#  fi
+#}
```
Comment (plaidman) [inline projects/L4T/packages/libglvnd/package.mk:34]: probably delete the commented code if it's not needed

Comment (plaidman) [reply projects/L4T/packages/libglvnd/package.mk:34]: there's a bunch of places where there's commented code. would it be annoying or easier for you if I pointed them all out?

File: projects/L4T/packages/xorg-server/package.mk (hunk 7 above)
Comment (GavinDarkglider) [reply projects/L4T/packages/xorg-server/package.mk:31]: I just copied the newer version from upstream libreelec. and added some extra patches to fix building against libglvnd. I plan to clean up the whole tree, once I rework the BSP/Kernel stuff. Have plans to make it more configurable and easier for people porting other L4T platforms to lakka. When that is done, we will probably have a lot of changes in the L4T specific packages.

File: projects/L4T/devices/Switch/packages/moonlight/package.mk (hunk 8 above)
Comment (GavinDarkglider) [reply projects/L4T/devices/Switch/packages/moonlight/package.mk]: This build is very L4T specific, and was not intended for anything other than the switch/jetson-nano. Putting it with other libretro cores and setting that flag to aarch64 would build and install a broken core on all arm64 boards. Safer to keep this here, and add support for this branch to main tree, but it only supports x86_64 - https://github.com/rock88/moonlight_libretro. Might also need makefile for linux added.

File: projects/L4T/options (hunk 9 above)
Comment (GavinDarkglider) [reply projects/L4T/options:102]: Yes, this can be removed.

File: projects/L4T/packages/libglvnd/package.mk (hunk 10 above)
Comment (GavinDarkglider) [reply projects/L4T/packages/libglvnd/package.mk:34]: I plan on cleaning all of this up at some point, I am just lazy, and am trying to get basic support in upstream builds.

File: projects/L4T/packages/tegra-bsp/package.mk (hunk 5 above)
Comment (GavinDarkglider) [reply projects/L4T/packages/tegra-bsp/package.mk:176]: Yeah, sorry, I copied that from a different old branch, from pre merging the switch stuff upstream, this file is actually an old version from one of my old vulkan testing branches with a lot of minor modifications to fix decoder stuff, and switch to proper glvnd, as that is how nvidia does it in ubuntu. Also, egl libs being set up properly is required for the decoder, so I didnt have much choice but to change the bsp package, same can be said with the other

File: projects/L4T/devices/Switch/packages/moonlight/package.mk (hunk 3 above)
Comment (GavinDarkglider) [reply projects/L4T/devices/Switch/packages/moonlight/package.mk]: Actually, this version is very switch specific, and should not be moved into the main tree. There is a whole other repo for non switch moonlight core, but this one was patched upstream for lakka switch, and old version at that. Plaidman and I just got it to build in tree, and I then applied a patch to it to force it to use nvmpi decoder over software decoder.
